### PR TITLE
Nextcloud Talk service

### PR DESCRIPTION
**Pull request by bertrandda**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to [the community](https://community.gladysassistant.com/) for testing before merging.
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.js`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Please provide a description of the change here. It's always best with screenshots, so don't hesitate to add some!

Added [Nextcloud Talk](https://nextcloud.com/talk/) integration to talk with Gladys throught Nextcloud Talk app.

### CODE REVIEW

File: front/src/routes/integration/all/nextcloud-talk/actions.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    store.setState({
+      nextcloudTalkGetSettingsStatus: NextcloudTalkStatus.Getting
+    });
+
+    let nextcloudUrl = '';
+    let nextcloudBotUsername = '';
+    let nextcloudBotPassword = '';
+    let nextcloudBotToken = '';
+
+    store.setState({
```
Comment (atrovato) [inline front/src/routes/integration/all/nextcloud-talk/actions.js:30]: Please call `store.setState` once, merging `nextcloudTalkGetSettingsStatus` and `nextcloud` vars

File: front/src/utils/consts.js [hunk 2]
```diff
@@ -94,6 +94,12 @@ export const GetWeatherModes = {
   DailyForecast: 'dailyForecast'
 };
 
+export const NextcloudTalkStatus = {
```
Comment (atrovato) [inline front/src/utils/consts.js]: Not sure we want `NextcloudTalkStatus` in global `consts.js` file.

File: server/lib/user/index.js [hunk 3]
```diff
@@ -8,6 +8,8 @@ const { getBySelector } = require('./user.getBySelector');
 const { getPicture } = require('./user.getPicture');
 const { getUserCount } = require('./user.getUserCount');
 const { getByTelegramUserId } = require('./user.getByTelegramUserId');
+const { getByNextcloudTalkToken } = require('./user.getByNextcloudTalkToken');
+const { getNextcloudTalkTokens } = require('./user.getNextcloudTalkTokens');
```
Comment (atrovato) [inline server/lib/user/index.js]: Not we want Nextcloud as a part of Gladys core.
It should be a service, not changing Gladys core.

File: server/services/nextcloud-talk/lib/message.bot.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+class NextcloudTalkBot extends EventEmitter {
+  constructor(gladys, serviceId, token) {
+    super();
+    this.gladys = gladys;
+    this.serviceId = serviceId;
+    this.token = token;
+    this.isPolling = false;
+  }
+
+  async startPolling() {
```
Comment (atrovato) [inline server/services/nextcloud-talk/lib/message.bot.js]: We usually extract class methods in single files.

**Review by atrovato — CHANGES_REQUESTED** [review]
I stop reading when I see you add a "nextcloud" token in Gladys core user model / table.
I am quite sure this is not the way to do this.
You may use service variables... or let's talk about point together.

**Comment by VonOx** [comment]
@atrovato this is the way telegram is implemented, no?

**Comment by atrovato** [comment]
@VonOx @Pierre-Gilles 
I don't think it's a good way to do.
I don't think we want to add information in user table each time we add a communication service.

**Comment by bertrandda** [comment]
As @VonOx says, it's work like that with Telegram. I tried to reproduce something structurally close for Nextcloud Talk to maintain code more easily for both services.
In Telegram service there is a `telegram_user_id` in user model, it's used here https://github.com/GladysAssistant/Gladys/blob/dd2b5f0bd37ca6f7ce6ca2fa51c7df032462d8b1/server/lib/message/message.sendToUser.js#L35-L38 to know where send message for a specific user

File: server/services/nextcloud-talk/lib/message.bot.js (hunk 4 above)
Comment (bertrandda) [reply server/services/nextcloud-talk/lib/message.bot.js]: I think it's a bit particular, we will need to handle multiple bots in the same time, one for each user and they can have different status. If you have a better way to do it we can still change it

File: front/src/utils/consts.js (hunk 2 above)
Comment (bertrandda) [reply front/src/utils/consts.js]: I am removing `NextcloudTalkStatus` and changing to existing `RequestStatus`

**Comment by Pierre-Gilles** [comment]
Yeah about the Telegram implementation, I'm not sure we want to do this for all new services, I agree this was my mistake :) 

What about just using the variable with a user_id parameter? 

You can create variable in the t_variable table, it's scoped by service & by user so you can have a special key to store this value, one by user, and no need to change the database schema for this (it's key/value)

**Comment by bertrandda** [comment]
Thank you @atrovato & @Pierre-Gilles. So I change the way to store the token. It now use variable (same as for username, password & url).
In Gladys core, I juste added `variable.getVariables.js` to fetch all existing variables for a same name and start correctly all user's Nextcloud bots. I also added variable lib inside `MessageHandler` to be able to retrieve token and send message through Nextcloud Talk
Is it better for you ?

**Comment by bertrandda** [comment]
Not stale, ready for reviews

File: server/services/nextcloud-talk/lib/message/message.bot.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
+const EventEmitter = require('events');
+const get = require('get-value');
+const logger = require('../../../../utils/logger');
+
+class NextcloudTalkBot extends EventEmitter {
```
Comment (Pierre-Gilles) [inline server/services/nextcloud-talk/lib/message/message.bot.js]: We don't really use this pattern in Gladys (class with inheritance in one big file)

To stay consistent with the rest of the project, could you do it with a prototype and 1 function per file like on other services?

File: server/services/nextcloud-talk/lib/message/message.disconnect.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/**
+ * @description Disconnect Nextcloud Talk
+ * @returns {Promise} - Resolve.
+ * @example
+ * disconnect();
+ */
+async function disconnect() {
+  if (this.bots.length > 0) {
+    Promise.all(this.bots.map((bot) => bot.stopPolling()));
```
Comment (Pierre-Gilles) [inline server/services/nextcloud-talk/lib/message/message.disconnect.js]: Isn't that too aggressive? Use Promise.each

File: server/services/nextcloud-talk/lib/message/message.bot.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    if (
+      lastKnownMessageId &&
+      newMessages &&
+      newMessages[newMessages.length - 1].actorId !== this.NEXTCLOUD_BOT_USERNAME
+    ) {
+      this.emit('message', newMessages.pop());
+    }
+
+    if (this.isPolling) {
+      setTimeout(() => this.poll(result.headers['x-chat-last-given'] || lastKnownMessageId), 50);
```
Comment (Pierre-Gilles) [inline server/services/nextcloud-talk/lib/message/message.bot.js]: Are you polling every 50ms forever? It's a bit extreme no?

File: server/lib/message/message.sendToUser.js [hunk 8]
```diff
@@ -30,11 +30,26 @@ async function sendToUser(userSelector, text, file = null) {
     payload: messageCreated,
   });
   // We send the message to the telegram service
-  const service = this.service.getService('telegram');
+  const telgramService = this.service.getService('telegram');
   // if the service exist and the user had telegram configured
-  if (service && user.telegram_user_id) {
+  if (telgramService && user.telegram_user_id) {
```
Comment (Pierre-Gilles) [inline server/lib/message/message.sendToUser.js]: Typo

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for this PR ! A few feedbacks :)

File: server/services/nextcloud-talk/lib/message/message.bot.js (hunk 5 above)
Comment (bertrandda) [reply server/services/nextcloud-talk/lib/message/message.bot.js]: I rewrite full process, it should be better

File: server/services/nextcloud-talk/lib/message/message.bot.js (hunk 7 above)
Comment (bertrandda) [reply server/services/nextcloud-talk/lib/message/message.bot.js]: It's not really that, it requests for new message here https://github.com/GladysAssistant/Gladys/pull/1388/files#diff-388fb960051d373b1c5591fe8a071c088ff028c5d4628956265ebbbadc6b6b3bR16-R20
There is a timeout define in the request query parameters to 15, the request end on new message response OR on timeout, then retry if we polling is enable, I added a delay of 50ms to possibly reduce load but it's not really necessary.

Comment (Pierre-Gilles) [reply server/services/nextcloud-talk/lib/message/message.bot.js]: But ` this.bots[userId].isPolling = true` right ?

So it's always doing a request, wait 50ms, one other request?

Considering the request takes 100ms, it means we are requesting every 150ms forever

It means 6.6 req/seconds = 2160 req/min = 129k requests per hour = 3 millions requests per day = 90 millions req/month

It's wayyy to much 😅

Comment (Pierre-Gilles) [reply server/services/nextcloud-talk/lib/message/message.bot.js]: Ok makes sense, I went reading the code and indeed!

File: front/src/config/i18n/en.json [hunk 9]
```diff
@@ -319,6 +319,22 @@
...
+      "introduction": "To speak through Nextcloud Talk, you need a specific Nextcloud account for your bot.",
+      "url": "Nextcloud host",
+      "urlInfo": "Enter your Nextcloud base URL.",
+      "username": "Username Nextcloud bot account",
+      "usernameInfo": "Enter Nextcloud bot account username.",
+      "password": "Password Nextcloud bot account",
+      "passwordInfo": "Enter Nextcloud bot account password.",
+      "token": "Bot conversation token",
+      "tokenInfo": "Enter conversation token between your Nextcloud account and the bot account.",
+      "configurationError": "Failed to save configration.",
```
Comment (Pierre-Gilles) [inline front/src/config/i18n/en.json]: *configuration

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thank you for making the necessary changes, I appreciate it the PR is really nice now 🙏

Code is clean, and well designed. Great work 👏

I spotted a super small typo, but nothing really serious.

If the PR is properly working for users, I'm good to merge! 👍

File: server/services/nextcloud-talk/index.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+   * @public
+   * @description This function stops the Nextcloud Talk Service
+   * @example
+   * gladys.services.nextcloudTalk.stop();
+   */
+  async function stop() {
+    logger.info('stopping Nextcloud Talk service');
+    await messageHandler.disconnect();
+  }
+
```
Comment (Pierre-Gilles) [inline server/services/nextcloud-talk/index.js:35]: Oh, and just one other feedback, can you implement the new function `isUsed` ?

It should return true/false if the user is actively using the service :) 

Like: 

https://github.com/GladysAssistant/Gladys/blob/master/server/services/zigbee2mqtt/index.js#L38

File: server/services/nextcloud-talk/lib/message/message.connect.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * @example
+ * connect([
+ *  { value: 'token1', user_id: 'f0de00a8-8ba7-4a4e-8f5a-7a21e94f36a8' },
+ *  { value: 'token2', user_id: '7f53d6ff-714f-4f08-a17b-df361a746d81' },
+ * ]);
+ */
+async function connect(tokens) {
+  // if a bot exist, we disconnect
+  if (Object.keys(this.bots).length > 0) {
+    await this.disconnect();
```
Comment (Pierre-Gilles) [inline server/services/nextcloud-talk/lib/message/message.connect.js:15]: This use case seems not tested according to codecov?

File: server/test/services/nextcloud-talk/index.test.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
+const { expect } = require('chai');
+const NextcloudTalkService = require('../../../services/nextcloud-talk');
+
+describe('nextcloud-talk', () => {
+  it('should start service', async () => {
+    const gladys = {
+      variable: {
+        getVariables: () => Promise.resolve([{ user_id: 'f0de00a8-8ba7-4a4e-8f5a-7a21e94f36a8', value: 'testToken' }]),
+        getValue: () => Promise.resolve('testvalue'),
```
Comment (Pierre-Gilles) [inline server/test/services/nextcloud-talk/index.test.js]: Why don't you use fake for this?

fake.resolve() 

?

It let's you test that the function was indeed called

Example:

https://github.com/GladysAssistant/Gladys/blob/ee812a2afae3aeeabce870d3e607975a7a026d63/server/test/services/mqtt/lib/installContainer.test.js#L81

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Thanks for your changes, it very nice! :) Great PR 🙏 

I have some super small comments but nothing too serious.

File: server/services/nextcloud-talk/lib/message/message.connect.js (hunk 11 above)
Comment (bertrandda) [reply server/services/nextcloud-talk/lib/message/message.connect.js:15]: It should be covered now

**Comment by Pierre-Gilles** [comment]
@bertrandda Is this ready for review ? Ping me if yes :)

**Comment by bertrandda** [comment]
@Pierre-Gilles yes code is ready, just need user real life tests

**Comment by Pierre-Gilles** [comment]
@bertrandda Was there some real-life tests since then?

**Comment by bertrandda** [comment]
> @bertrandda Was there some real-life tests since then?

Except mine no new tests

**Comment by Pierre-Gilles** [comment]
@bertrandda Ok, if we don't have news on the forum in 1 week, and you tested carefully on your side, I'm ok to merge it to master :) We can't wait forever for testers ! (Don't hesitate to ping me if I forget)

**Comment by Pierre-Gilles** [comment]
@bertrandda Sorry I did a few Eslint changes that needs to be applied to this PR

I'm ok to merge after, no need to wait for more testers :)

**Pull request merged**